Focus the document on the POPRF construction and its properties (#284)
cfrg/draft-irtf-cfrg-voprf: draft-irtf-cfrg-voprf.md

--- abstract

An Oblivious Pseudorandom Function (OPRF) is a two-party protocol between
client and server for computing the output of a Pseudorandom Function (PRF).
The server provides the PRF secret key, and the client provides the PRF
input. At the end of the protocol, the client learns the PRF output without
learning anything about the PRF secret key, and the server learns neither
the PRF input nor output. A Partially-Oblivious PRF (POPRF) is an OPRF

> wherein both client and server also agree on some public input to the

Author: ```suggestion
that allows client and server to provide public input to the
```

PRF. OPRFs and POPRFs can also satisfy a notion of 'verifiability'.
In this setting, clients verify that the server's output is the result of

> evaluating the underlying PRF with an expected a public key. This document

Author: Much better!

- Server: Computes the pseudorandom function over a private key. Learns
  nothing about the client's input.
- NIZK: Non-interactive zero knowledge.
- DLEQ: Discrete Logarithm Equality.

# POPRF Protocol {#protocol}

In this section, we define two POPRF variants: a base mode and verifiable
mode. In the base mode, a client and server interact to compute
`output = F(skS, input, info)`, where `input` is the client's private input,
`skS` is the server's private key, `info` is the public input (or metadata),
and `output` is the POPRF output. The client learns `output` and the server
learns nothing. In the verifiable mode, the client also receives proof that
the server used `skS` in computing the function.

To achieve verifiability, as in the original work of {{JKK14}}, we
provide a zero-knowledge proof that the key provided as input by the
server in the `Evaluate` function is the same key as it used to produce
their public key. As an example of the nature of attacks that this
prevents, this ensures that the server uses the same private key for
computing the verifiable POPRF output and does not attempt to "tag"
individual clients with select keys. This proof does not reveal the

> server's long-term private key to the client.

Author: 👍

## Context APIs {#api}

In this section, we detail the APIs available on the client and server
POPRF contexts. Each API has the following implicit parameters:

- GG, a prime-order group implementing the API described in {{pog}}.
- contextString, a domain separation tag constructed during context setup.

The data types `PrivateInput` and `PublicInput` are opaque byte strings
of arbitrary length no larger than 2^13 octets. `Proof` is a concatenated
sequence of two `SerializedScalar` values, as shown below.

~~~
struct {
  SerializedScalar c;
  SerializedScalar s;
} Proof;
~~~

### Server Context

The ServerContext encapsulates the context string constructed during

> setup and the POPRF key pair. It has three functions, `Evaluate`,

Author: Keeping as-is.

  c = GG.HashToScalar(h2Input)
  s = (r - c * k) mod p

>   proof = GG.SerializeScalar(c) || GG.SerializeScalar(s)

Reviewer: Proof is a struct, isn't it?

Author: I don't think these are incompatible -- this line says "create a Proof struct that contains serialized c and s". Please suggest something concrete if you think this should be done differently.

## Error Considerations

Some POPRF APIs specified in this document are fallible. For example, `Finalize`
and `Evaluate` can fail if any element received from the peer fails deserialization.
The explicit errors generated throughout this specification, along with the
conditions that lead to each error, are as follows:

> - `VerifyError`: Verifiable POPRF proof verification failed; {{verifiable-unblind}}.

Reviewer: do we need to include an error when the inverse of `k+m` does not exist?

Author: Probably. File a new issue?

This section discusses the cryptographic security of our protocol, along
with some suggestions and trade-offs that arise from the implementation
of a POPRF.

## Security Properties {#properties}

> The security properties of an OPRF protocol with functionality

Reviewer: ```suggestion
The security properties of a POPRF protocol with functionality
```

non-malleable with high probability, and so a pseudorandom function must
be non-malleable to maintain indistinguishability.

A POPRF protocol must also satisfy the following property:

- Partial obliviousness: The server must learn nothing about the client's
  private input or the output of the function. In addition, the client must
  learn nothing about the server's private key. Both client and server learn
  the public input (info).

Essentially, partial obliviousness tells us that, even if the server learns
the client's private input x at some point in the future, then the server will
not be able to link any particular POPRF evaluation to x. This property is
also known as unlinkability {{DGSTV18}}.

Optionally, for any protocol that satisfies the above properties, there
is an additional security property:

> - Verifiable: The client must only complete execution of the protocol if

Reviewer: ```suggestion
- Verifiability: The client must only complete execution of the protocol if
```

Reviewer: hmm, not so sure.

  it can successfully assert that the POPRF output it computes is
  correct. This is taken with respect to the POPRF key held by the
  server.

> Any POPRF that satisfies the 'verifiable' security property is known as a

Reviewer: ```suggestion
Any POPRF that satisfies the 'verifiability' property is known as a
```

Author: Keeping as-is.

verifiable POPRF. In practice, the notion of verifiability requires that
the server commits to the key before the actual protocol execution takes

> place. Then the client verifies that the server has used the key in the
> protocol using this commitment. In the following, we may also refer to this

Reviewer: ```suggestion
place. Then the client uses this commitment to verify that the server has used the key in the
protocol. In the following, we may also refer to this
```

commitment as a public key.

## Cryptographic Security {#cryptanalysis}

Below, we discuss the cryptographic security of the verifiable POPRF
protocol from {{protocol}}, relative to the necessary cryptographic
assumptions that need to be made.

### Protocol Security and Computational Hardness Assumptions {#assumptions}

The POPRF construction in this document is based on the construction known
as 3HashSDHI given by {{TCRSTW21}}. The construction is identical to
3HashSDHI, except that this design can optionally perform multiple POPRF
evaluations in one go, whilst only constructing one NIZK proof object.
This is enabled using an established batching technique.

The cryptographic security of the construction is based on the assumption

> that the One-More Gap SDHI assumption from {{TCRSTW21}} is computationally

Reviewer: One-More Gap Strong Diffie-Hellman Inversion (SDHI) assumption

Reviewer: Computational Diffie-Hellman (CDH) assumption

Author: 👍

difficult to solve. {{TCRSTW21}} show that both the One-More Gap CDH
assumption and the One-More Gap SDHI assumption reduce to the q-DL assumption
in the algebraic group model, for some q number of `Evaluate` queries.
(The One-More Gap CDH assumption was the hardness assumption used to
evaluate the 2HashDH-NIZK construction from {{JKK14}}, which is a predecessor
to the design in this specification.)

### Static q-DL Assumption

A side-effect of the POPRF design is that it allows instantiation of an oracle for

> retrieving "strong-DH" evaluations, in which an adversary can query a group element

Reviewer: double check if this is accurate
`retrieving instances of the strong Diffie-Hellman problem, in which an adversary can query a group element`

Blinding may also be done additively. However, the choice of blinding
mechanism has security implications. {{JKX21}} analyze the security
properties of different blinding mechanisms. The results can be
summarized as follows:

- Multiplicative blinding is safe for all applications.
- Additive blinding is possibly unsafe, unless one of the following conditions
  are met:
    - The client has a certified copy of the server public key (as is the case
      in the verifiable mode);
    - The client input has high entropy; and
    - The client mixes the public key into the OPRF evaluation.

To avoid security issues, where some of the above conditions may not be met,
this specification use of multiplicative blinding. This is because it is
not known if the server public key is available or if the client input has
high entropy.

## Timing Leaks

To ensure no information is leaked during protocol execution, all
operations that use secret data MUST run in constant time. Operations that
SHOULD run in constant time include all prime-order group operations and
proof-specific operations (`GenerateProof()` and `VerifyProof()`).

> ## Key Rotation {#key-rotation}

Reviewer: does the first paragraph still applies to POPRF?
